rfd/0203-database-healthchecks.md

```diff
@@ -0,0 +1,606 @@
+---
+authors: Gavin Frazar (gavin.frazar@goteleport.com)
+state: draft
+---
+
+# RFD 0203 - Database Health Checks
+
+## Required Approvals
+
+- Engineering: @r0mant && @greedy52
+- Product: @roraback
+
+## What
+
+Teleport database agents will periodically perform health checks to a database's endpoint and report the results as a health status in the agent's `db_server` heartbeat.
+
+When a Teleport proxy routes a user connection, the proxy will prioritize a `healthy` agent->DB status over "less healthy" agent->DB status if multiple agents proxy the same database.
+
+## Why
+
+The primary motivation for this change is to enable smarter proxy->agent routing when multiple DB agents heartbeat the same database (i.e., an HA setup [^1]).
+
+[^1]: https://goteleport.com/docs/enroll-resources/database-access/guides/ha/#combined-replicas
+
+Currently, in an HA DB agent configuration, the proxy will randomly choose one of the DB agents to handle a user connection.
+
+If a subset of the DB agents can't reach the database endpoint, then user connection attempts will randomly fail.
+
+To fix this, the Teleport proxy service can sort the DB agents by target health status and preferentially dial agents that have healthy network connectivity to the database.
+
+Secondary motivations for this change:
```

Comment:
I honestly think that motivations are inverted, or at least equally important :) UX should always be primary motivation IMO.